<a href="https://colab.research.google.com/github/BigBubl1k/data_science/blob/main/Yandex_market_%D0%A2%D1%8E%D1%80%D0%B8%D0%BD_%D0%98%D0%B3%D0%BE%D1%80%D1%8C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#импортируем библиотеки и загрузим данные в data_json

In [69]:
from urllib.request import urlopen
import json
url = 'http://5.181.109.139/marketer/api/data.json'
response = urlopen(url)
data_json = json.loads(response.read())


#Проверим типы данных и посмотрим на их запись

In [70]:
len(data_json)

2

In [71]:
type(data_json)

dict

In [72]:
type(data_json['items'])

list

In [73]:
data_json['items'][0]

{'competitor price': 1799,
 'competitor shop': 'VeloSmile',
 'original category': 'Спасательные жилеты и круги',
 'original id': '2094028',
 'original price': 3210,
 'original title': '"Cпасательный жилет ""Докер"" до 80 кг р.44-46"',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=aprice',
 'sima-land.ru price': 1955,
 'sima-land.ru updated': '2022-04-19 01:00:10'}

#Сделаем цикл, который проходится по каждому объекту словаря

Первый цикл пройдем вручную

In [52]:
article = data_json['items'][0]['original id']                      #артикул товара
whole_sale_price = data_json['items'][0]['sima-land.ru price']      #оптовая цена этого товара
original_price = data_json['items'][0]['original price']            #наша цена до преобразования
min_competitor_price = data_json['items'][0]['competitor price']    #минимальная цена конкурента на этот товар
our_price = (50+whole_sale_price)*100/55                            #расчет по формуле нашей цены товара

if our_price < min_competitor_price:                                #если наша цена окажется ниже миинимальной цены конкурентов, устанавливаем ее
  data_json['items'][item]['original price'] = min_competitor_price*0.97
data_json['items'][0]

{'competitor price': 1799,
 'competitor shop': 'VeloSmile',
 'original category': 'Спасательные жилеты и круги',
 'original id': '2094028',
 'original price': 3210,
 'original title': '"Cпасательный жилет ""Докер"" до 80 кг р.44-46"',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=aprice',
 'sima-land.ru price': 1955,
 'sima-land.ru updated': '2022-04-19 01:00:10'}

Теперь пройдемся по всем элементам

In [53]:
for item in data_json['items']:                       #проходимся по каждому объекту

  if item['competitor shop'] == 'Русэкспресс':        #если самая низкая цена конкурента от Русэкспресса,
    continue                                          #то пропускаем шаг цикла, т.к это наш магазин


  whole_sale_price = item['sima-land.ru price']       #оптовая цена товара
  min_competitor_price = item['competitor price']     #минимальная цена конкурента на этот товар


  if item['original id'] == article:                  #Так как нам нужен только первый набор данных по артиклу, 
    item['original price'] = original_price
    continue                                          #если артикул повторяет артикул прошлого цикла, пропускаем шаг

  else:
    article = item['original id']                     #если артикул новый, то переменной article присваем новое значение артикула
    our_price = (50+whole_sale_price)*100/55          #и считаем нашу цену по формуле

    original_price = item['original price']           #чтобы для других наборов данных с одинаковым артикулом также изменить
                                                      #цену на новую, если изменения будут, используем переменную original_price

  if our_price < min_competitor_price:                #если эта цена окажется ниже мин. цены конкурента, заменяем в данных ее
      item['original price'] = min_competitor_price*0.97
      original_price = min_competitor_price*0.97

  else:
    continue




In [54]:
#327100 записей всего

In [55]:
type(data_json['items'][0]['competitor price'])

int

In [56]:
data_json['items'][6]

{'competitor price': 1084,
 'competitor shop': 'Qwen',
 'original category': 'Игрушечные машинки и техника',
 'original id': '7517061',
 'original price': 1051.48,
 'original title': '100 секретных окошек "Транспорт',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/1699704981/offers?lr=213&sku=101625914255&how=aprice',
 'sima-land.ru price': 518,
 'sima-land.ru updated': '2022-04-19 01:00:11'}

In [57]:
data_json['items'][7]

{'competitor price': 1084,
 'competitor shop': 'Qwen',
 'original category': 'Игрушечные машинки и техника',
 'original id': '7517061',
 'original price': 1051.48,
 'original title': '100 секретных окошек "Транспорт',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/1699704981/offers?lr=213&sku=101625914255&how=aprice',
 'sima-land.ru price': 518,
 'sima-land.ru updated': '2022-04-19 01:00:11'}

Проверим, что для тех наборов данных, где мы изменили цену, цена изменилась для всех наборов с тем же артикулом

In [58]:
print('original price:',data_json['items'][6]['original price'],';','original price:',data_json['items'][7]['original price'])

original price: 1051.48 ; original price: 1051.48


In [59]:
for item in data_json['items'][6]:
  print(item ,type(item))

original id <class 'str'>
original title <class 'str'>
original price <class 'str'>
original category <class 'str'>
original url <class 'str'>
original updated <class 'str'>
sima-land.ru price <class 'str'>
sima-land.ru updated <class 'str'>
competitor shop <class 'str'>
competitor price <class 'str'>


Как видно по строке original price, цена скорретировалась (была изначально 1051), т.е, за счет того, что наша цена оказалась изначально ниже самой низкой цены конкурента, нам удалось поднять цену на товар на 48 коп.

#Теперь сформируем xml файл

In [60]:
def json2xml(json_obj, line_padding=""):
    result_list = list()

    json_obj_type = type(json_obj)

    if json_obj_type is list:
        for sub_elem in json_obj:
            result_list.append(json2xml(sub_elem, line_padding))

        return "\n".join(result_list)

    if json_obj_type is dict:
        for tag_name in json_obj:
            sub_obj = json_obj[tag_name]
            result_list.append("%s<%s>" % (line_padding, tag_name))
            result_list.append(json2xml(sub_obj, "\t" + line_padding))
            result_list.append("%s</%s>" % (line_padding, tag_name))

        return "\n".join(result_list)

    return "%s%s" % (line_padding, json_obj)

In [61]:
result = json2xml(data_json)
print(result[:5000])

<items>
	<original id>
		2094028
	</original id>
	<original title>
		"Cпасательный жилет ""Докер"" до 80 кг р.44-46"
	</original title>
	<original price>
		3210
	</original price>
	<original category>
		Спасательные жилеты и круги
	</original category>
	<original url>
		https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=aprice
	</original url>
	<original updated>
		2022-04-18 00:37:18
	</original updated>
	<sima-land.ru price>
		1955
	</sima-land.ru price>
	<sima-land.ru updated>
		2022-04-19 01:00:10
	</sima-land.ru updated>
	<competitor shop>
		VeloSmile
	</competitor shop>
	<competitor price>
		1799
	</competitor price>
	<original id>
		2094028
	</original id>
	<original title>
		"Cпасательный жилет ""Докер"" до 80 кг р.44-46"
	</original title>
	<original price>
		3210
	</original price>
	<original category>
		Спасательные жилеты и круги
	</original category>
	<original url>
		https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=a

In [62]:
print(result[-150:])

ed>
	<competitor shop>
		kv.mart
	</competitor shop>
	<competitor price>
		319
	</competitor price>
</items>
<updated>
	2022-04-20 06:00:08
</updated>


In [36]:
for i in data_json:
  print(i)

items
updated


In [45]:
data_json['updated']

'2022-04-20 06:00:08'

In [63]:
import xml.etree.cElementTree as ET

In [66]:
root = ET.Element('root')

#ET.SubElement(root, 'items').text = data_json['items']

items = ET.SubElement(root, "items")

for i in data_json['items']:

  ET.SubElement(items, 'original id').text = i['original id']
  ET.SubElement(items, 'original title').text = i['original title']
  ET.SubElement(items, 'original price').text = str(i['original price'])
  ET.SubElement(items, 'original category').text = i['original category']
  ET.SubElement(items, 'original url').text = i['original url']
  ET.SubElement(items, 'original updated').text = i['original updated']
  ET.SubElement(items, 'sima-land.ru price').text = str(i['sima-land.ru price'])
  ET.SubElement(items, 'sima-land.ru updated').text = i['sima-land.ru updated']
  ET.SubElement(items, 'competitor shop').text = i['competitor shop']
  ET.SubElement(items, 'competitor price').text = str(i['competitor price'])

updated = ET.SubElement(root, "updated").text = data_json['updated']

tree = ET.ElementTree(root)
tree.write('yafbs_prices_updated1.xml')


In [67]:
type(tree)


xml.etree.ElementTree.ElementTree

In [68]:
import xml.etree.ElementTree as ET
from xml.etree.ElementTree import tostring
tree=ET.parse('yafbs_prices_updated1.xml')
tree = tree.getroot()
t = tostring(tree)
t = t.lower()
tree= ET.fromstring(t)

ParseError: ignored

In [61]:
with open("yafbs_prices_updated.xml", "w") as f:
    f.write(ET.tostring(tree))

AttributeError: ignored

In [59]:
import xml.etree.ElementTree as ET
 
# Passing the path of the
# xml document to enable the
# parsing process
tree = ET.parse('/content/yafbs_prices_updated.xml')
 
# getting the parent tag of
# the xml document
root = tree.getroot()
 
# printing the root (parent) tag
# of the xml document, along with
# its memory location
print(root)
 
# printing the attributes of the
# first tag from the parent
print(root[0].attrib)
 
# printing the text contained within
# first subtag of the 5th tag from
# the parent
print(root[5][0].text)

ParseError: ignored